node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+			}
+		})
+		.collect::<Vec<PeerId>>();
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::SendMessage(
+			interested_peers,
+			BitfieldDistribution::PROTOCOL_ID,
+			Encode::encode(&message),
+		),
+	))
+	.await?;
+	Ok(())
+}
+
+/// Handle an incoming message from a peer.
+async fn process_incoming_peer_message<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	origin: PeerId,
+	message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	// we don't care about this, not part of our view
+	if !tracker.view.contains(&message.relay_parent) {
+		return modify_reputation(ctx, origin, COST_NOT_INTERESTED).await;
+	}
+
+	// Ignore anything the overseer did not tell this subsystem to work on
+	let mut job_data = tracker.per_relay_parent.get_mut(&message.relay_parent);
+	let job_data: &mut _ = if let Some(ref mut job_data) = job_data {
+		job_data
+	} else {
+		return modify_reputation(ctx, origin, COST_NOT_INTERESTED).await;
+	};
+
+	let validator_set = &job_data.validator_set;
+	if validator_set.len() == 0 {
+		return modify_reputation(ctx, origin, COST_MISSING_PEER_SESSION_KEY).await;
+	}
+
+	// use the (untrusted) validator index provided by the signed payload
+	// and see if that one actually signed the availability bitset
+	let signing_context = job_data.signing_context.clone();
+	let validator_index = message.signed_availability.validator_index() as usize;
+	let validator = if let Some(validator) = validator_set.get(validator_index) {
+		validator.clone()
+	} else {
+		return modify_reputation(ctx, origin, COST_VALIDATOR_INDEX_INVALID).await;
+	};
+
+	if message
+		.signed_availability
+		.check_signature(&signing_context, &validator)
+		.is_ok()
+	{
+		let one_per_validator = &mut (job_data.one_per_validator);
+		// only relay_message a message of a validator once
+		if one_per_validator.get(&validator).is_some() {
+			trace!(target: "bitd", "Alrady received a message for validator at index {}", validator_index);
+			return Ok(());
+		}
+		one_per_validator.insert(validator.clone(), message.clone());
+
+		relay_message(ctx, job_data, &mut tracker.peer_views, validator, message).await
+	} else {
+		return modify_reputation(ctx, origin, COST_SIGNATURE_INVALID).await;
+	}
+}
+
+/// Deal with network bridge updates and track what needs to be tracked
+/// which depends on the message type received.
+async fn handle_network_msg<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	bridge_message: NetworkBridgeEvent,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	match bridge_message {
+		NetworkBridgeEvent::PeerConnected(peerid, _role) => {
+			// insert if none already present
+			tracker.peer_views.entry(peerid).or_insert(View::default());
+		}
+		NetworkBridgeEvent::PeerDisconnected(peerid) => {
+			// get rid of superfluous data
+			tracker.peer_views.remove(&peerid);
```

Comment:
The `PeerId` will still be present in the `per_relay_parent` data. Better to clean up, although I can think of some rationalizations why not to.

Reply:
Unfortunately we can not clean them up, if a peer decides to disconnect and reconnect we would lose the information of which messages we already sent ( + received) and as such that would cause the node to lose reputation ( or be subject to flooding ).